### PR TITLE
SQS Ops Scripts

### PR DESCRIPTION
**Pull request by danielmarbach**
Connects to https://github.com/Particular/PlatformDevelopment/issues/1108

### CODE REVIEW

**Comment by danielmarbach** [comment]
@Particular/aws-maintainers here is the current WIP. Early feedback is already welcome.

File: Snippets/Sqs/Sqs_All/QueueDeletion/Tests.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    [TestFixture]
+    [Explicit]
+    public class Tests
+    {
+
+        [Test]
+        public void DeleteAllQueues()
+        {
+            QueueDeletionUtils.DeleteAllQueues().GetAwaiter().GetResult();
```
Comment (SimonCropp) [inline Snippets/Sqs/Sqs_All/QueueDeletion/Tests.cs]: i dont think `.GetAwaiter().GetResult();` is required here?

File: Snippets/Sqs/Sqs_All/Sqs_All.csproj [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net462</TargetFramework>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="ApprovalTests" Version="3.*" />
```
Comment (SimonCropp) [inline Snippets/Sqs/Sqs_All/Sqs_All.csproj]: do u need ApprovalTests?

File: Snippets/Sqs/Sqs_All/Sqs_All.csproj [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    <PackageReference Include="Mono.Cecil" Version="0.*" />
```
Comment (SimonCropp) [inline Snippets/Sqs/Sqs_All/Sqs_All.csproj]: same goes for cecil

File: Snippets/Sqs/Sqs_All/Sqs_All.csproj [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    <PackageReference Include="NServiceBus" Version="6.*" />
+    <PackageReference Include="NServiceBus.AmazonSQS" Version="3.0.0-alpharelease0057" />
+    <PackageReference Include="NUnit" Version="3.*" />
+    <PackageReference Include="System.Management.Automation" Version="6.*" />
+    <ProjectReference Include="..\..\Common\Common.csproj" />
+    <Reference Include="System.Messaging" />
```
Comment (SimonCropp) [inline Snippets/Sqs/Sqs_All/Sqs_All.csproj]: suspect this got pulled over from msmq?

Comment (danielmarbach) [inline Snippets/Sqs/Sqs_All/Sqs_All.csproj]: yes, I planned to remove stuff I don't need. Not done yet

File: Snippets/Sqs/Sqs_All/QueueDeletion/Tests.cs (hunk 1 above)
Comment (danielmarbach) [inline Snippets/Sqs/Sqs_All/QueueDeletion/Tests.cs]: No, need to fix this

**Comment by SimonCropp** [comment]
looks good so far

File: Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                {
+                    QueueUrl = queueUrl,
+                    WaitTimeSeconds = 1,
+                    MaxNumberOfMessages = 10,
+                })
+                .ConfigureAwait(false);
+
+            if (receivedMessages.Messages.Count == 0)
+            {
+                return;
```
Comment (ahofman) [inline Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs]: Is this safe - are we sure that if we long-poll for 1 second that we will definitely not miss any messages that may be still in the queue? Would it be safer to long-poll for 20 seconds?

File: Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+                    continue;
+                }
+
+                var failedQueueName = ReadFailedQueueHeader(message);
+                var failedQueueUrlResponse = await client.GetQueueUrlAsync(failedQueueName)
+                    .ConfigureAwait(false);
+                var failedQueueUrl = failedQueueUrlResponse.QueueUrl;
+                // Large message don't need to be handled separately since the S3BodyKey is preserved
+                await client.SendMessageAsync(new SendMessageRequest(failedQueueUrl, message.Body)) // what to do with the attributes?
+                    .ConfigureAwait(false);
```
Comment (ahofman) [inline Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs]: When the pump moves poison messages to the error queue, it doesn't move SQS message attributes with it, I wouldn't worry about the attributes for now

File: Snippets/Sqs/Sqs_All/QueueCreation/QueueCreationUtils.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+                    {
+                        QueueName = QueueNameHelper.GetSqsQueueName(queueName)
+                    };
+                    var value = Convert.ToInt32((maxTimeToLive ?? DefaultTimeToLive).TotalSeconds).ToString();
+                    sqsRequest.Attributes.Add(QueueAttributeName.MessageRetentionPeriod, value);
+                    await client.CreateQueueAsync(sqsRequest)
+                        .ConfigureAwait(false);
+                }
+            }
+            catch (QueueNameExistsException)
```
Comment (ahofman) [inline Snippets/Sqs/Sqs_All/QueueCreation/QueueCreationUtils.cs]: When a queue is created with the same name as another queue that was deleted less than 60sec ago, `CreateQueueAsync` throws `QueueDeletedRecentlyException` - the user would just need to wait 60 sec and try again. Do we want to add this handling here?

File: Snippets/Sqs/Sqs_All/QueueCreation/QueueCreation.ps1 [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+)
+
+# main queue
+CreateQueue -QueueName $EndpointName -MaxTimeToLive $MaxTimeToLive
+
+# timeout queue
+CreateQueue -QueueName "$EndpointName.timeouts" -MaxTimeToLive $MaxTimeToLive
+
+# timeout dispatcher queue
+CreateQueue -QueueName "$EndpointName.timeoutsdispatcher" -MaxTimeToLive $MaxTimeToLive
```
Comment (ahofman) [inline Snippets/Sqs/Sqs_All/QueueCreation/QueueCreation.ps1]: Important to be aware that SQS queue names are case-sensitive (unlike most other transports) - is this the same case that endpoints expect?

File: Snippets/Sqs/Sqs_All/QueueDeletion/QueueDeletionUtils.cs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            using (var client = ClientFactory.CreateSqsClient())
+            {
+                await DeleteQueues(client)
+                    .ConfigureAwait(false);
+            }
+        }
+
+        static async Task DeleteQueues(IAmazonSQS client)
+        {
+            var queuesToDelete = await client.ListQueuesAsync("")
```
Comment (ahofman) [inline Snippets/Sqs/Sqs_All/QueueDeletion/QueueDeletionUtils.cs]: This deletes every queue in the region, may be a bit heavy-handed :) The queue name prefix is typically used to isolate application environments in the same region (e.g., "DEV", "STAGING", "PROD", etc). 

Perhaps allow the queue name prefix to be passed in as a parameter?

(Unfortunately this suggestion cross-cuts pretty much everything in this PR - sorry :P)

File: Snippets/Sqs/Sqs_All/QueueCreation/QueueCreation.ps1 (hunk 8 above)
Comment (danielmarbach) [inline Snippets/Sqs/Sqs_All/QueueCreation/QueueCreation.ps1]: Good point. No the core uses

https://github.com/Particular/NServiceBus/blob/926db772c885d729591352713809d40267d18bc3/src/NServiceBus.Core/DelayedDelivery/TimeoutManager/TimeoutManager.cs#L66 and

https://github.com/Particular/NServiceBus/blob/926db772c885d729591352713809d40267d18bc3/src/NServiceBus.Core/DelayedDelivery/TimeoutManager/TimeoutManager.cs#L86

Will change. Good catch

Comment (danielmarbach) [inline Snippets/Sqs/Sqs_All/QueueCreation/QueueCreation.ps1]: I added a test which spins up endpoints

File: Snippets/Sqs/Sqs_All/QueueDeletion/QueueDeletionUtils.cs (hunk 9 above)
Comment (danielmarbach) [inline Snippets/Sqs/Sqs_All/QueueDeletion/QueueDeletionUtils.cs]: I don't think it cross cuts everything. Most other scripts just accept the queue name or the endpoint name. Do you agree that it is enough to just add a prefix param to the deletion scripts?

File: Snippets/Sqs/Sqs_All/QueueCreation/QueueCreationUtils.cs (hunk 7 above)
Comment (danielmarbach) [inline Snippets/Sqs/Sqs_All/QueueCreation/QueueCreationUtils.cs]: Not sure if it is required. I think for the queue creation it is OK to bomb in that case since we cannot operate with the queues not present but we can leave with queues still there in the worst case right?

Comment (danielmarbach) [inline Snippets/Sqs/Sqs_All/QueueCreation/QueueCreationUtils.cs]: Never mind I get what you are saying. Do we also need to deal with the truncation?

Comment (ahofman) [inline Snippets/Sqs/Sqs_All/QueueCreation/QueueCreationUtils.cs]: I'm not sure about truncation - I added that to the transport as an internal-only configuration option solely to get the ATs working, I'm not sure if it's a good way to handle real systems. To date I have just been relying on users choosing endpoint names that are short enough - this gives a very straightforward experience when a user goes looking for their queue in the SQS portal. I'm a little concerned that the pretruncation will be more confusing for new users and will give a worse experience compared to just getting an exception saying "queue name is too long", but I'm open to suggestions here.

In any case I don't think we should add the pretruncation support here because it's not possible for users to configure it in their endpoints at this time.

File: Snippets/Sqs/Sqs_All/QueueCreation/QueueCreationUtils.cs [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+    #region sqs-create-queues
+
+    public static class QueueCreationUtils
+    {
+        static TimeSpan DefaultTimeToLive = TimeSpan.FromDays(4);
+
+        public static async Task CreateQueue(string queueName, TimeSpan? maxTimeToLive = null, string queueNamePrefix = null, bool preTruncateQueueNames = false)
+        {
+            try
```
Comment (SimonCropp) [inline Snippets/Sqs/Sqs_All/QueueCreation/QueueCreationUtils.cs]: should this try be narrowed to only surround the CreateQueueAsync?

File: Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+                    MaxNumberOfMessages = 10,
+                })
+                .ConfigureAwait(false);
+
+            if (receivedMessages.Messages.Count == 0)
+            {
+                return;
+            }
+
+            foreach (var message in receivedMessages.Messages)
```
Comment (SimonCropp) [inline Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs]: minor suggestion. i find this more readable 

            var foundMessage = receivedMessages.Messages
                .SingleOrDefault(message => message.MessageId != messageId);

            if (foundMessage != null)
            {
                var failedQueueName = ReadFailedQueueHeader(foundMessage);
                var failedQueueUrlResponse = await client.GetQueueUrlAsync(failedQueueName)
                    .ConfigureAwait(false);
                var failedQueueUrl = failedQueueUrlResponse.QueueUrl;
                // Large message don't need to be handled separately since the S3BodyKey is preserved
                await client.SendMessageAsync(new SendMessageRequest(failedQueueUrl, foundMessage.Body)) // what to do with the attributes?
                    .ConfigureAwait(false);
                await client.DeleteMessageAsync(queueUrl, foundMessage.ReceiptHandle)
                    .ConfigureAwait(false);
                return;
            }

File: Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+            await InspectMessagesUntilFound(client, messageId, queueUrl)
+                .ConfigureAwait(false);
+        }
+
+        static string ReadFailedQueueHeader(Message message)
+        {
+            var headers = ExtractHeaders(message);
+            var queueName = headers.Single(x => x.Key == "NServiceBus.FailedQ").Value;
+            return QueueNameHelper.GetSqsQueueName($"{queueName}");
```
Comment (SimonCropp) [inline Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs]: shouldnt `$"{queueName}"` be `queueName`?

**Comment by SimonCropp** [comment]
overall very nice work :)

File: Snippets/Sqs/Sqs_All/QueueCreation/QueueCreationUtils.cs (hunk 7 above)
Comment (SimonCropp) [inline Snippets/Sqs/Sqs_All/QueueCreation/QueueCreationUtils.cs]: i generally dont like adding truncation. they are hard for the user to reason about. IMO it is not something we should be adding abstraction logic to

**Comment by danielmarbach** [comment]
Ok so I incorporate all the feedback. From a code perspective, this is done. Will now start covering all the snippets in the markdown.

**Comment by danielmarbach** [comment]
@Particular/aws-maintainers this is done from my perspective.

File: Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs (hunk 5 above)
Comment (danielmarbach) [inline Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs]: I changed the long poll to 20 secs

File: Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+                // Large message don't need to be handled separately since the S3BodyKey is preserved
+                await client.SendMessageAsync(new SendMessageRequest(failedQueueUrl, foundMessage.Body)) // what to do with the attributes?
+                    .ConfigureAwait(false);
+                await client.DeleteMessageAsync(queueUrl, foundMessage.ReceiptHandle)
+                    .ConfigureAwait(false);
+
+                return;
+            }
+
+            await InspectMessagesUntilFound(client, messageId, queueUrl)
```
Comment (ahofman) [inline Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs]: This could be difficult to use if the user wants to return many messages to the source queue because this receives other messages in the error queue and marks them as invisible. The user may get in a loop of run script for 1 message -> wait for the visibility timeout to expire (20sec?) -> run script for next message -> wait for timeout -> etc, which would really hurt for 00's of messages. Is this something to worry about?

File: Snippets/Sqs/Sqs_All/QueueNameHelper.cs [hunk 14]
```diff
@@ -29,8 +29,8 @@ public static string GetSqsQueueName(string destination, string queueNamePrefix
 
             if (s.Length > 80)
             {
-                throw new Exception(
-                    $"Address {destination} with configured prefix {queueNamePrefix} is longer than 80 characters and therefore cannot be used to create an SQS queue. Use a shorter queue name.");
+                throw new ArgumentException(
```
Comment (danielmarbach) [inline Snippets/Sqs/Sqs_All/QueueNameHelper.cs]: These changes should be ported to the PowerShell version as well (currently copy pasted into each PS file)

File: Snippets/Sqs/Sqs_All/QueueCreation/CreateEndpointQueues.cs [hunk 15]
```diff
@@ -24,18 +24,21 @@ static class CreateEndpointQueues
 
         public static async Task CreateQueuesForEndpoint(string endpointName, TimeSpan? maxTimeToLive = null, string queueNamePrefix = null, bool includeRetries = false)
         {
+            // timeout dispatcher queue
+            // This queue is created first because it has the longest name. 
+            // If the endpointName and queueNamePrefix are too long this call will throw and no queues will be created. 
+            // In this event, a shorter value for endpointName or queueNamePrefix should be used.
+            await QueueCreationUtils.CreateQueue($"{endpointName}.TimeoutsDispatcher", maxTimeToLive, queueNamePrefix)
```
Comment (danielmarbach) [inline Snippets/Sqs/Sqs_All/QueueCreation/CreateEndpointQueues.cs]: 👍

File: Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs (hunk 13 above)
Comment (danielmarbach) [inline Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs]: How would you mitigate this? I originally had the wait timeout pretty low but that has other implications

Comment (SimonCropp) [inline Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs]: I think a warning in the .md will suffice

Comment (ahofman) [inline Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs]: I'd suggest changing the method to accept a collection of message ids instead of just one.

At a glance it looks like the [RabbitMQ scripts](https://docs.particular.net/transports/rabbitmq/operations-scripting#return-message-to-source-queue-the-retry-helper-methods) have a similar issue.

Are these scripts intended to be a starting point / reference / example only or are they meant to be bullet-proof for production usage as-is? If the former I'd say leave it and add a warning per @SimonCropp suggestion.

File: Snippets/Sqs/Sqs_All/QueueNameHelper.cs (hunk 14 above)
Comment (ahofman) [inline Snippets/Sqs/Sqs_All/QueueNameHelper.cs]: Done

File: Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs (hunk 13 above)
Comment (SimonCropp) [inline Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs]: @ahofman they are intended as a starting point. how "bullet-proof" we make them depends on how much effort it takes and how much more difficult it makes the snippets to understand/explain. in the long run i would like to see this stuff packaged up into helper libs/nugets. but that is outside the scope of the current work

Comment (ahofman) [inline Snippets/Sqs/Sqs_All/ErrorQueue/ErrorQueue.cs]: Cool good to know :) I've added a warning to the .md.

File: Snippets/Sqs/Sqs_All/QueueCreation/QueueCreation.ps1 [hunk 16]
```diff
@@ -125,8 +128,8 @@ Add-Type @'
 
             if (s.Length > 80)
             {
-                throw new Exception(
-                    string.Format("Address {0} with configured prefix {1} is longer than 80 characters and therefore cannot be used to create an SQS queue. Use a shorter queue name.", destination, queueNamePrefix));
+                throw new ArgumentException(
```
Comment (danielmarbach) [inline Snippets/Sqs/Sqs_All/QueueCreation/QueueCreation.ps1]: This has to be changed in all ps files :(

**Comment by danielmarbach** [comment]
@SimonCropp  https://github.com/Particular/docs.particular.net/pull/3042#discussion_r132893704 has not been addressed. Will correct it on master tomorrow
